patches/server/0814-Overridable-YAML-configuration.patch

```diff
@@ -0,0 +1,330 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Cubxity <cubxity@protonmail.com>
+Date: Sun, 3 Oct 2021 08:03:16 +0200
+Subject: [PATCH] Overridable YAML configuration
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index f421e6a2e43e0a673dbb8a9a2b4331387e523e02..702477b94b28f0463fa51ee2aa999515457eaa68 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -18,6 +18,7 @@ import java.util.logging.Level;
+ import java.util.regex.Pattern;
+ 
+ import com.google.common.collect.Lists;
++import io.papermc.paper.configuration.OverridableYamlConfiguration;
+ import net.minecraft.server.MinecraftServer;
+ import org.bukkit.Bukkit;
+ import org.bukkit.ChatColor;
+@@ -55,7 +56,7 @@ public class PaperConfig {
+ 
+     public static void init(File configFile) {
+         CONFIG_FILE = configFile;
+-        config = new YamlConfiguration();
++        config = new OverridableYamlConfiguration("paper");
+         try {
+             config.load(CONFIG_FILE);
+         } catch (IOException ex) {
+diff --git a/src/main/java/io/papermc/paper/configuration/EnvironmentOverrideSource.java b/src/main/java/io/papermc/paper/configuration/EnvironmentOverrideSource.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..3cd6579bb79bdda501a5ee33c04fd5a21fa015aa
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/configuration/EnvironmentOverrideSource.java
+@@ -0,0 +1,30 @@
++package io.papermc.paper.configuration;
++
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.Locale;
++
++/**
++ * An implementation of {@link OverrideSource}, which fetches values from system's environment.
++ * The requested path will be transformed in the following manner:
++ * - dot (.) will be replaced with an underscore (_),
++ * - dash (-) will be replaced with an underscore (_),
++ * - and it will be capitalized.
++ * <p>
++ * The group name will be prepended, followed by an underscore (_).
++ */
++public final class EnvironmentOverrideSource implements OverrideSource {
```

Comment:
Is there actually any reason to implement these in their own classes ?
Looking at the `OverrideSource` interface, it seems to be a pretty straight forward candidate for a `FunctionalInterface`.
E.g. instead of having to mess around with private constructors and instances, 

You could move these as constants into the `OverrideSource` interface, e.g.
```java
/**
 * An implementation of {@link OverrideSource}, which fetches values from system's environment.
 * The requested path will be transformed in the following manner:
 * - dot (.) will be replaced with an underscore (_),
 * - dash (-) will be replaced with an underscore (_),
 * - and it will be capitalized.
 * <p>
 * The group name will be prepended, followed by an underscore (_).
 */
OverrideSource ENVIRONMENT_SOURCE = (path, group) -> {
    final var environmentKey = group + "_" + path.replace(".", "_").replace("-","_");
    return System.getenv(environmentKey.toUpperCase(Locale.ROOT));
};
```